Fix System.IO.Ports.SerialPort GetPortNames() Returns garbage chars on Win10IoT Devices with USB Serial ports

PRIMETSS opened the pull request:
Issue:
https://github.com/dotnet/corefx/issues/35722
https://github.com/dotnet/corefx/issues/32847

On Windows 10 IoT device (eg Rasbery Pi 3+) with Serial USB device connected `System.IO.Ports.SerialPort .GetPortNames() `returns garbage non chars.
![image](https://user-images.githubusercontent.com/11331370/62270410-29626b80-b475-11e9-857c-e0129a6bb905.png)

Reason:
`System.IO.Ports.SerialPort .GetPortNames()' was reading registry key 
![image](https://user-images.githubusercontent.com/11331370/62270526-56af1980-b475-11e9-8243-f483073ba092.png)
But this appears not to be correctly initialised by Serial Function Driver (serenum.sys) when USB device is connect to a Win10IoT device.
[https://docs.microsoft.com/en-us/windows-hardware/drivers/serports/external-naming-of-com-ports](https://docs.microsoft.com/en-us/windows-hardware/drivers/serports/external-naming-of-com-ports)
 
Suggested Fix:
Since there appears to be no sure fire way to enumerate both legacy Com & USB PnP COM ports on Windows OS, and Win10 IoT also seem to handle initialisation of ports differently again.
Suggested returning both the Registry COM[x] port names and also the DosDevice HardwareID (?) Port Name to allow connection to previously un-able to open ports due to no valid port name to open.

Hardware Tests:
As testing this issue requires multiple hardware, not only different OS's but also a variety of different COM port devices, eg Legacy on board UARTS IC's and also USB PnP Serial Ports. Testing is laborious.

Tested: 
Legacy Desktop System (Intel MB, on-board Serial port UART) (Win7) [On-board Serial Port + Multiple USB PnP comm port devices]
Raspberry Pi 3B+ (Win10 IoT ARM32) [Multiple USB PnP comm port devices + OnBoard SoC UART]
HP Pro Desktop (Windows 10 Pro 64 bit) [Multiple USB PnP comm port devices]

Windows 8 { Not tested yet }

PRIMETSS committed:
Fixing issue with Win10IoT on Raspberry Pi was not able to connect to USB Serial port devices, as usbser.sys was writing non ascii characters to registery key HKLM\HARDWARE\DEVICEMAP\SERIALCOMM

Multiple approaches where tried, but opted for adding a non breaking method (GetPortDosDeviceNames() to obtain raw DosDevice names by adding a Interop method to call kernel32.dll!QueryDosDevice()

This way consumer can use either method that suits depending on application.

Ref Issue: https://github.com/dotnet/corefx/issues/18023

Known Issues: On Desktop PC's  kernal32.dll!QueryDosDevice() lists BOTH COM4 and also \\?\USB####... ID for same COM port. i.e. duplicate.
Couldnt find a suitable way to remove this duplice that didnt involve heavy Registry traversal that might not be reliable as registering COM ports seems not that clear/fragmented.
Thought best was to add new method GetPortDosDeviceNames() and alllow consumer posibility of removing duplcate anmes (as duplicate port name would still connect)

PRIMETSS committed:
Fixed spaces and API
Removed GetPortDosDeviceNames() method that would require API  change  review.
Still need to handle checking for running on Win10IoT there issue observed, if more testing proves duplicate ports an issue?
Removed Marshal.AllocHGlobal(maxSize) in QueryDosDeviceComPorts(string guid)
But need check to see if this is ok?

PRIMETSS committed:
Refactore QueryDosDeviceComPorts() & GetPortNames()

QueryDosDeviceComPorts(): Remove old comments from prev code using // Marshal.AllocHGlobal(maxSize); for memBuffer

GetPortNames(): Imporoved code to filter corrupt chars in Win10IoT

PRIMETSS committed:
Update System.IO.Ports.csproj

Update <Compile Include="$(CommonPath)\Interop\Windows\Kernel32\Interop.QueryDosDevice.cs">
Conditions

watfordgnf commented on src/System.IO.Ports/src/System.IO.Ports.csproj:
```diff
@@ -129,7 +129,7 @@
     <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Kernel32\Interop.FormatMessage.cs">
       <Link>Common\CoreLib\Interop\Windows\Interop.FormatMessage.cs</Link>
-    </Compile>
+    </Compile>   
```
```suggestion
    </Compile>
```

PRIMETSS committed:
Added Support for UAP

Needed to add overload support for base SerialStream.OpenPort() to SerialStream.UAP.
Failing Builds for UAP platform

PRIMETSS committed:
Relocate include for Interop.QueryDosDevice.cs

Causing Build errors on UAP

PRIMETSS committed:
Update System.IO.Ports.csproj

Test to find build issue

PRIMETSS committed:
Revert "Update System.IO.Ports.csproj"

This reverts commit 0660e1282d799ed21930e0657114f19ebf57c771.

PRIMETSS committed:
Update PinvokeAnalyzerExceptionList.analyzerdata.netstandard

Added kernel32.dll!QueryDosDevice to baseline

PRIMETSS committed:
Update SerialPort.Win32.cs

Check for running On Win10IoT to include DosDevicves as they are triggering Test Fails which are difficult to refactor

PRIMETSS committed:
Fix to Failing Tests running in CI

System.IO.Ports.SerialPort.OpenDevices was failing becuase it expected an exeption on all found non serial port, however some did open! So included checks for a valid portname in test and also when opening the SerialPort.Open()

Also added conditional check to either/or use DosDevices names only if running on Win10IoT (Which needs improvement)

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -13,26 +15,111 @@ public static string[] GetPortNames()
             //
-            // WMI: https://msdn.microsoft.com/en-us/library/aa394413.aspx
-            // QueryDosDevice: https://msdn.microsoft.com/en-us/library/windows/desktop/aa365461.aspx
+            // WMI: https://msdn.microsoft.com/en-us/library/aa394413.aspx [But Requires .NetFramework !]
```
```suggestion
            // WMI: https://msdn.microsoft.com/en-us/library/aa394413.aspx [requires .NetFramework]
```

krwq replied:
seems this didn't get fixed

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -13,26 +15,111 @@ public static string[] GetPortNames()
-            // WMI: https://msdn.microsoft.com/en-us/library/aa394413.aspx
-            // QueryDosDevice: https://msdn.microsoft.com/en-us/library/windows/desktop/aa365461.aspx
+            // WMI: https://msdn.microsoft.com/en-us/library/aa394413.aspx [But Requires .NetFramework !]
+            // QueryDosDevice: https://msdn.microsoft.com/en-us/library/windows/desktop/aa365461.aspx [Can produces duplicates eg COM4 also listed as DosDeviceName \\USB#VID_1B4F&PID_214F..]
```
```suggestion
            // QueryDosDevice: https://msdn.microsoft.com/en-us/library/windows/desktop/aa365461.aspx [can produce duplicates eg COM4 also listed as DosDeviceName \\USB#VID_1B4F&PID_214F..]
```

krwq replied:
didn't get fixed

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -13,26 +15,111 @@ public static string[] GetPortNames()
+            // This causes issues with System.IO.Ports.Tests
+            // Port detection broken on Windows IoT (Does not initialise regitry with COM port names), so use QueryDosDevice
+            //
+            // ToDO: Be good to Limit this to Win10IoT only, not just check for ARM arch!!
```
remove the TODO, we will deal with it when there are any issues

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -13,26 +15,111 @@ public static string[] GetPortNames()
+                // Typical DosDevice serial port name 'USB#VID_1B4F&PID_214F&MI_00#6&381731fd&0&0000#{86e0d1e0-8089-11d0-9ce4-08003e301f73}'
+                //
+                foreach (string dosName in QueryDosDeviceComPorts("86e0d1e0-8089-11d0-9ce4-08003e301f73"))
+                    resultPortNames.Add(dosName.ToLower());
```
braces around foreach body

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -13,26 +15,111 @@ public static string[] GetPortNames()
+            HashSet<string> resultPortNames = new HashSet<string>();
+
+            // If running on Windows IoT, search and add the Serial Devices from QueryDosDevice
+            // If this is allowed to happen on Windows devices where port detection happens normally, duplicate Serial Port names are added.
```
will all devices be duplicated? if so should we just not call the other API at all when running on ARM? If not is there some easy way to remove duplicates?

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -13,26 +15,111 @@ public static string[] GetPortNames()
+            // Build devices matching guid for serial ports
+            foreach (string name in allDevicesArray)
+            {
+                if (name.ToLower().Contains(guid))
```
string.Compare with OrdinalIgnoreCase (same and elsewhere with similar pattern)

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.cs:
```diff
@@ -643,7 +648,8 @@ public int Read(byte[] buffer, int offset, int count)
                 if (bytesReadToBuffer == count)
                 {
-                    if (_readPos == _readLen) _readPos = _readLen = 0;  // just a check to see if we can reset buffer
+                    if (_readPos == _readLen)
```
braces around if body (lack of braces are only allowed for argument validation or if all ifs don't use braces)

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialStream.Uap.cs:
```diff
@@ -15,5 +15,12 @@ public SafeFileHandle OpenPort(uint portNumber)
                 Interop.Kernel32.GenericOperations.GENERIC_READ | Interop.Kernel32.GenericOperations.GENERIC_WRITE,
                 Interop.Kernel32.FileOperations.FILE_FLAG_OVERLAPPED);
         }
+        public SafeFileHandle OpenPort(string portName)
```
nit: enter before

krwq commented on src/System.IO.Ports/tests/SerialPort/OpenDevices.cs:
```diff
@@ -19,9 +19,12 @@ public void OpenDevices01()
             foreach (KeyValuePair<string, string> keyValuePair in dosDevices)
             {
-                if (!string.IsNullOrEmpty(keyValuePair.Key) && !comPortNameRegex.IsMatch(keyValuePair.Key))
+                if (!string.IsNullOrEmpty(keyValuePair.Key) && !comPortNameRegex.IsMatch(keyValuePair.Key) && !keyValuePair.Key.Contains("86e0d1e0-8089-11d0-9ce4-08003e301f73"))
```
this guid should be a constant somewhere instead of being inline everywhere (constant should have a comment where it's coming from)

PRIMETSS replied:
Thanks for input!
Yes agree!! But struggling trying to find a suitable location to place the Constant....
Any suggestions?

krwq replied:
anywhere, just make it internal. For tests you can define a duplicate constant

PRIMETSS committed:
Update src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs

Co-Authored-By: Krzysztof Wicher <mordotymoja@gmail.com>

PRIMETSS committed:
Update src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs

Co-Authored-By: Krzysztof Wicher <mordotymoja@gmail.com>

PRIMETSS committed:
Fixes to PR

PRIMETSS committed:
Merge branch 'primetss/system.io.ports' of https://github.com/PRIMETSS/corefx into primetss/system.io.ports

PRIMETSS committed:
Revert "Merge branch 'primetss/system.io.ports' of https://github.com/PRIMETSS/corefx into primetss/system.io.ports"

This reverts commit a7ae441814b86d4a3a785ae5eb017ae8da0c2c0f, reversing
changes made to a5dbf0dc58b194ffb706cd64ceb2dd82864f45fa.

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -11,6 +11,8 @@ namespace System.IO.Ports
 {
     public partial class SerialPort : Component
     {
+        const string GUID_DEVINTERFACE_COMPORT = "86e0d1e0-8089-11d0-9ce4-08003e301f73"; // SerialPort GUID Class ID
```
nit: explicit `private`
nit2: comment above the line
nit3: if the constant is not defined elsewhere with this exact casing please use UsualCasing

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -3,36 +3,125 @@
+                //
+                foreach (string dosName in QueryDosDeviceComPorts(GUID_DEVINTERFACE_COMPORT))
+                {
+                    resultPortNames.Add(dosName.ToLower());
```
ToLowerInvariant or if possible use OrdinalIgnoreCase comparer for HashSet

PRIMETSS replied:
Ok sure, Ta!
Thanks for your help appreciated!

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -3,36 +3,125 @@
+                    if (serialKey != null)
+                        foreach (string valueName in serialKey.GetValueNames())
+                        {
+                            string portName = ((string)serialKey.GetValue(valueName)).ToUpper();
```
please use comparison with ignore case and use invariant culture instead of creating new copy of string

PRIMETSS replied:
Am fixing these as i go, but just dawned on me is Culture an issue in regards to Device Names?
As there are a lot of string comparisons done in the tests that are not Culture aware.
Also if user was to change language of PC, would this not mean registry values would have to be re-written?

JeremyKuhne replied:
I don't believe these are ever localized.

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -3,36 +3,125 @@
+
+                            // Add the key's value (portname) if it is valid (Win10IoT places garbage in this Reg Key)
+                            if (portName.StartsWith("COM")) // Filters corrupt chars for COM[x] devices in Registry on Win10IoT 
+                                resultPortNames.Add(portName);
```
nit: braces

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -3,36 +3,125 @@
+            }
+
+            string[] result = new string[resultPortNames.Count];
+            resultPortNames.CopyTo(result);
```
`ToArray()`? (System.Linq)

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -3,36 +3,125 @@
-                    }
-                    return result;
+                    byte[] memBuff = new byte[maxSize];
+                    fixed (byte* memPtr = memBuff)
```
please use braces here

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -3,36 +3,125 @@
-                    return result;
+                    byte[] memBuff = new byte[maxSize];
+                    fixed (byte* memPtr = memBuff)
+                        try
```
is the try..finally here just to avoid braces?

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.cs:
```diff
@@ -598,6 +598,11 @@ public void Open()
+            // Check its a valid Serial Port name
+            if (!_portName.Contains("COM") && !_portName.Contains("86e0d1e0-8089-11d0-9ce4-08003e301f73"))
+                throw new ArgumentException(SR.Format(SR.Arg_InvalidSerialPort, _portName), nameof(_portName));
+
```
nit: double enter

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.cs:
```diff
@@ -598,6 +598,11 @@ public void Open()
                 throw new InvalidOperationException(SR.Port_already_open);
 
+            // Check its a valid Serial Port name
+            if (!_portName.Contains("COM") && !_portName.Contains("86e0d1e0-8089-11d0-9ce4-08003e301f73"))
```
please replace hard-coded value with constant

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.cs:
```diff
@@ -643,13 +648,18 @@ public int Read(byte[] buffer, int offset, int count)
-                    if (_readPos == _readLen) _readPos = _readLen = 0;  // just a check to see if we can reset buffer
+                    if (_readPos == _readLen)
+                    {
+                        _readPos = _readLen = 0;  // just a check to see if we can reset buffer
```
nit: comment above

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialStream.Windows.cs:
```diff
@@ -336,14 +336,17 @@ public override int ReadTimeout
 
-                if (constant == infiniteTimeoutConst) return SerialPort.InfiniteTimeout;
-                else return constant;
+                if (constant == infiniteTimeoutConst)
```
`return (constant == infiniteTimeoutConst) ? SerialPort.InfiniteTimeout : constant;`

PRIMETSS committed:
Fixes

Remove hard coded GUID const, code tidy and removed string case conversions

krwq commented:
Seems you got some merge conflicts in this PR you'll need to resolve

PRIMETSS committed:
Merge branch 'master' into primetss/system.io.ports

PRIMETSS committed:
Fixed Spaces

PRIMETSS committed:
Update OpenDevices.cs

Rename GuidDevInterfaceComPort camelCCase

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -13,26 +16,111 @@ public static string[] GetPortNames()
+            return resultPortNames.ToArray();
+        }
+
+        private static string[] QueryDosDeviceComPorts(string filterGuid)
```
return `IEnumerable<string>` and no ToArray in the end - this way you don't have to copy data multiple times

PRIMETSS replied:
Changing the return type is a API change though isnt it?

JeremyKuhne replied:
On the public methods, yes. But you can add the individual results directly to the output if you `yield` them out of here (no need to create a separate array).

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.cs:
```diff
@@ -596,6 +598,10 @@ public void Open()
                 throw new InvalidOperationException(SR.Port_already_open);
 
+            // Check its a valid Serial Port name
+            if (!_portName.Contains("COM") && !_portName.Contains("dev/tty") && !_portName.Contains(GuidDevInterfaceComPort))
```
I'm not sure this change is appropriate here - @wfurt, any opinions? I'd vote to remove this check in general and let the native API throw

JeremyKuhne replied:
I'd remove the checks- it's what we're doing as a general policy now with `System.IO`. Run the port name through `Path.GetFullPath()` as I've suggested elsewhere. That will throw only for embedded nulls, which you'd want anyway. It will also put the `\\.\` on `COMx` so that addition will need to be removed.

PRIMETSS replied:
Do we also need to worry about culture and case here.
Pretty sure in Windows all Serial ports get names "COMx" in upper case, but not sure about Linux device names, assumed they are all lower case..
I agree with @krwq and just let the Native API throw, as this check might actually stop a valid PortName from being able to be used. Which is worse!

carlossanlop replied:
I think architecture-specific details like COM vs dev/tty should be separated in their own separate implementations.
Since `SerialPort` is a partial class, I suggest you wrap this condition in a new method defined both in `SerialPort.Windows.cs` and also in a new file `SerialPort.Unix.cs`, then call it here.

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialStream.Win32.cs:
```diff
@@ -18,5 +18,15 @@ public SafeFileHandle OpenPort(uint portNumber)
+
+        public SafeFileHandle OpenPort(string portName)
+        {
+            // portName: eg "\\?\USB#VID_1B4F&PID_214F&MI_00#6&20b7e7e3&0&0000#{86e0d1e0-8089-11d0-9ce4-08003e301f73}"
```
If port name starts with `\\?` won't you end up with double `\\?` in the native call?

cc: @JeremyKuhne

JeremyKuhne replied:
We used to explicitly [disallow](https://referencesource.microsoft.com/#System/sys/system/io/ports/SerialPort.cs,417) `\\` but that wasn't necessary on any Windows that is remotely modern. We sort of still "block" it by putting the prefix on.

Aren't we no longer building UAP? If so, I'd fold this back in to the .Windows and get rid of the UAP'isms. If not, I'd change both to take "portName" and let the UAP code do the parsing for a number. I don't think we should be validating the names at all.

The right way to do this is call `Path.GetFullPath(portName)`, which will put the `\\.\` on `COM1`. Users should be specifying full device paths if they want direct access to other device names. Don't add the `\\?\` or `\\.\` here. We should give the full device name in the list of ports if they're not "legacy" `COMx` names.

PRIMETSS replied:
Sorry, double checked this, and my comment was wrong, included the preamble '\\?\'
But sound not have, Portnames retrived are just

> USB#VID_1B4F&PID_214F&MI_00#6&381731fd&0&0000#{86e0d1e0-8089-11d0-9ce4-08003e301f73}

krwq commented on src/System.IO.Ports/src/System/IO/Ports/SerialStream.Win32.cs:
```diff
@@ -18,5 +18,15 @@ public SafeFileHandle OpenPort(uint portNumber)
+                Interop.Kernel32.GenericOperations.GENERIC_READ | Interop.Kernel32.GenericOperations.GENERIC_WRITE,
+                0,              // comm devices must be opened w/exclusive-access
+                FileMode.Open,  // comm devices must use OPEN_EXISTING
+               Interop.Kernel32.FileOperations.FILE_FLAG_OVERLAPPED);
```
nit: indent off by 1

krwq commented:
Seems like few changes got reverted (some of the suggestions I've done before) - my main concern right now is prepending `\\?` which I'm not sure is correct

JeremyKuhne commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -13,26 +16,111 @@ public static string[] GetPortNames()
+            return resultPortNames.ToArray();
+        }
+
+        private static string[] QueryDosDeviceComPorts(string filterGuid)
```
This method is **very** costly. On my current machine I've got 798 devices. We should use `ArrayPool` for this and walk through the buffer with `ReadOnlySpan`. We'll have a convenience method for this eventually (#26528), but it isn't too terribly hard to do to write a loop to slice repeatedly at `IndexOf`.

JeremyKuhne replied:
The buffer needed to contain all of those is 54K. With the iterations here and marshaling I'd be surprised if we allocate less than a megabyte.

PRIMETSS replied:
Haven't mocked that mocked devices yet, but was the performance hit from Registry enumeration or HashSet?

Im questioning if we need to use HashSet here at all?
From testing, it seems that even if a duplicate COM port name exists (have to be user over ridden in devices [with ignoring warnings!]) the duplicate name does not get created in the registry (cant test Win7 yet)
Since a device name collision would be a badly configured system with un-expected results anyway, is it worth filtering for this unusual condition, or Im I missing something?
![image](https://user-images.githubusercontent.com/11331370/66246335-b9ad7d00-e752-11e9-9bdf-2f5c37a36c34.png)
![image](https://user-images.githubusercontent.com/11331370/66246348-d053d400-e752-11e9-9aa8-d66681af5076.png)
![image](https://user-images.githubusercontent.com/11331370/66246374-e2357700-e752-11e9-815e-83a6973c28e9.png)
![image](https://user-images.githubusercontent.com/11331370/66246395-ed88a280-e752-11e9-8811-a0ca1a76ce06.png)

PRIMETSS replied:
Im not great at this, but trying to diagnose this expensive function so I have a baseline to compare any changes to. (eg before going to try using ArrayPool etc)

On PC with >700 Registry serial port entries, appears the Interop call to Win32.Registry:Key:GetValue (reading the com port names) is taking the most?
How do to over come that? If this is where we need to read the info from?

Or, as there is more than one way to get com-port names (eg QueryDosDeviceComPorts() used to work around the garbage chars problem in Win10IoT) We just use that for all Windows targets? Even though it is still doing Interop calls, maybe its more efficient with its fill buffer in chunks/repeat approach. But I cant mock 700 real serial devices here!
![image](https://user-images.githubusercontent.com/11331370/66262997-cacdbb00-e832-11e9-842c-0c3f1c75ce7b.png)

JeremyKuhne commented on src/Common/src/Interop/Windows/Kernel32/Interop.QueryDosDevice.cs:
```diff
@@ -0,0 +1,16 @@
+    internal partial class Kernel32
+    {
+          [DllImport(Libraries.Kernel32, SetLastError = true)]
+          internal static extern uint QueryDosDevice(string lpDeviceName, IntPtr lpTargetPath, int ucchMax);
```
This should be `CharSet.Unicode` and ideally `QueryDosDeviceW` so that you can also set `ExactSpelling` to true in the `[DllImport]`.

PRIMETSS replied:
Ive added 

> ExactSpelling

Not sure about unicode? Would comm device names contain Unicode chars?
But guess QueryDosDeviceW exist for a reason?

JeremyKuhne replied:
It is a historical thing that .NET picks the "A" version and uses ANSI by default. It is almost never the right choice as .NET and Windows use the same Unicode encoding.

Windows stores all of this as Unicode, so when you call the "A" versions it's converting the Unicode to ANSI and then we're having to convert it back.

You can't use `ExactSpelling` unless the name actually matches. The DLL entry points are actually "QueryDosDeviceA" and "QueryDosDeviceW" for this API. The Windows headers have "QueryDosDevice" defined as one of those two depending on whether or not you've defined `UNICODE`. .NET probes for `A` or `W` names when it can't find the name you've given to make the process feel like writing C/C++ (there is no export called "QueryDosDevice"). We're preferring to be explicit in CoreFX to avoid accidentally using ANSI when we should almost always be using Unicode. (Like 99% of the time.)

PRIMETSS replied:
Interesting to find QueryDosDeviceW returns a double null terminated string unlike QueryDosDevice
Anyway, handled that ok I think.

krwq replied:
@PRIMETSS I suspect double 0 is because Unicode uses two bytes per character so it's actually one character which is 0 producing 2 bytes (although just a guess)

JeremyKuhne replied:
Yes, they both return a double null. It's a common Windows pattern for lists of strings. In ANSI it's `0x00 0x00` in bytes, in Unicode it's `0x00 0x00 0x00 0x00`.

JeremyKuhne commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -13,26 +16,111 @@ public static string[] GetPortNames()
+            uint returnSize = 0;
+            int maxSize = 1024;
+            string allDevices = null;
+            const uint ERROR_INSUFFICIENT_BUFFER = 122;
```
This is defined in https://github.com/dotnet/corefx/blob/master/src/Common/src/Interop/Windows/Interop.Errors.cs

JeremyKuhne commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -13,26 +16,111 @@ public static string[] GetPortNames()
+            string[] allDevicesArray = null;
+            List<string> returnList = new List<string>();
+
+            while (returnSize == 0)
```
The right way to do this loop is:

``` C#

while ((result = Interop.Kernel32.QueryDosDeviceW(deviceName, buffer, buffer.Length)) == 0)
{
    int error = Marshal.GetLastWin32Error();
    switch (error)
    {
        case Interop.Errors.ERROR_INSUFFICIENT_BUFFER:
            // Return and rent a larger char buffer
            break;
        default:
            throw new Win32Exception(error);
    }
}

// There are two trailing nulls, don't want an empty split
ReadOnlySpan<char> names = buffer.Slice(0, result - 2);

// Loop through and .Slice off each string, doing .Contains(guid, StringComparison.OrdinalIgnoreCase)

// Pass back matched names with `\\?\` prepended (as they're already "resolved")
```

PRIMETSS replied:
Struggling with this one, how to incorporate ArrayPool & unsafe pointer that needs to be re-assigned as each buffer expansion creates a new pointer reference..

krwq replied:
@PRIMETSS this will remind C++ manual memory management:

```csharp
byte[] array = ArrayPool.Rent(initialSize); // note size of array might be larger than requested

// too small
ArrayPool.Return(array);
array = ArrayPool.Rent(newSize);
```

+ making sure we return when something throws (i.e. try .. finally) and triple check you don't double return

JeremyKuhne replied:
You need to rent a buffer of `char`, not `byte`. The API `[DllImport]` signature can just take `char[]` for the buffer and it will be pinned, you have nothing else to do.

JeremyKuhne replied:
> making sure we return when something throws

If throws aren't expect to happen frequently (which is the case here) we don't need to return the buffer (and shouldn't). Adding a try block makes things slower and quickly eats away at the benefit of returning the buffer.

JeremyKuhne commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs:
```diff
@@ -13,26 +16,111 @@ public static string[] GetPortNames()
+                        returnSize = Interop.Kernel32.QueryDosDevice(null, (System.IntPtr)memPtr, maxSize);
+                        if (returnSize != 0)
+                        {
+                            allDevices = Marshal.PtrToStringAnsi((System.IntPtr)memPtr, (int)returnSize);
```
This won't be necessary with the Unicode version of the API.

PRIMETSS replied:
Not sure if better way to do it, still refactoring this part to handle the buffer sizing loop like you suggested.
Not sure if better way to convert the unicode , but ended doiing 
`allDevice = System.Text.Encoding.Unicode.GetString(memBuff)`
for now, This also handled the fact QueryDosDeviceW() returns a double null terminating string, unlike QueryDosDevice that does not. Wasn't expecting that

JeremyKuhne reviewed (changes requested):
Thanks for looking into this! I've left a few comments.

JeremyKuhne commented:
cc: @carlossanlop

PRIMETSS committed:
Suggested fixes

Work in progress still on how to handle SerialPort.Win32.cs Interop.Kernel32.QueryDosDeviceW() buffer expansion

PRIMETSS committed:
Relocate  GuidDevInterfaceComPort

Untill decision is made on checking for valid Serial Port Name in SerialPort.cs

PRIMETSS committed:
Refactor SerialPortWin32.cs QueryDosDeviceComPorts()

https://github.com/dotnet/corefx/pull/39943#discussion_r321537803

krwq commented on src/Common/src/Interop/Windows/Kernel32/Interop.QueryDosDevice.cs:
```diff
@@ -11,7 +11,10 @@ internal partial class Interop
-          [DllImport(Libraries.Kernel32, SetLastError = true, ExactSpelling = true, CharSet = CharSet.Unicode)]
-          internal static extern uint QueryDosDeviceW(string lpDeviceName, IntPtr lpTargetPath, int ucchMax);
+        [DllImport(Libraries.Kernel32, SetLastError = true, ExactSpelling = true, CharSet = CharSet.Unicode)]
+        internal static extern uint QueryDosDeviceW(string lpDeviceName, IntPtr lpTargetPath, int ucchMax);
```
is this overload used anywhere? I'd rather expose only `char*` overload than having two

PRIMETSS replied:
I would have thought it was used quite a few times, so I left it there for reference,
But to my surprise QueryDosDevice() isn't use that often.

Ive remove the one not in use.

PRIMETSS committed:
Update Interop.QueryDosDevice.cs

Remove un-used overload for QueryDosDevice()

https://github.com/dotnet/corefx/pull/39943#discussion_r322034162

carlossanlop commented on src/Common/src/Interop/Windows/Mincore/Interop.OpenCommPort.cs:
```diff
@@ -16,5 +16,11 @@ internal partial class mincore
+
+        [DllImport(Libraries.CoreComm_L1_1_1, SetLastError = true)]
+        internal static extern SafeFileHandle OpenCommPort(
+            string portName,  // Is Sybolic Device Name Possible here? eg '\\.\USB_VID......'
```
nit: typo:
```suggestion
            string portName,  // Is Symbolic Device Name Possible here? eg '\\.\USB_VID......'
```
Do you know what the answer is to this question? Can you please find out? I think it would be preferable having a comment stating if "it's possible" or "it's not possible", rather than an open question that may confuse future users of this method.

krwq commented:
@PRIMETSS is this ready for re-review please ping me if all comments are addressed

carlossanlop commented on src/System.IO.Ports/src/System/IO/Ports/SerialPort.cs:
```diff
@@ -784,7 +796,8 @@ private int InternalRead(char[] buffer, int offset, int count, int timeout, bool
             Debug.Assert(buffer.Length - offset >= count, "invalid offset/count!");
 
-            if (count == 0) return 0;   // immediately return on zero chars desired.  This simplifies things later.
+            if (count == 0)
```
Mentioning @krwq 's suggestion again for the ifs here and below:
> braces around if body (lack of braces are only allowed for argument validation or if all ifs don't use braces)

PRIMETSS committed:
Update src/Common/src/Interop/Windows/Mincore/Interop.OpenCommPort.cs

Co-Authored-By: Carlos Sanchez Lopez <1175054+carlossanlop@users.noreply.github.com>

PRIMETSS committed:
Fix Tests, And Interop OpenPorts issue

Fix: Issue with incorrect port name toInterop.Kernel32.CreateFile()
Fix:  Incorect Exception Type in System.IO.Ports\tests\SerialPort\PortName.PortName_COM257
Fix: System.IO.Ports\tests\Support\TCSupport.GenerateSerialInfo() Not reporting any hardware for system with 1 serial port
Fix: Typos & Comments

PRIMETSS committed:
Revert Detection of system with 1 COM port

// See Github issues #15961, #16033, #20764

Was testing with system with FTDI adapter and suspect as per issue #20764 FTDI hardware may be cause of invalid tmeouts

maryamariyan commented:
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your corefx repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/corefx <path to the patch created in step 1>`

krwq commented:
@PRIMETSS I need to close the PR because of the repo merge (need to close or merge all PRs before 5PM today and this one has a merge conflict).

Do you think you'll have time to fix this? If so let me know if you need help with resurrecting this in the new repo

PRIMETSS commented:
Im in a diff time zone and about to go to meeting for few hours.
But can look at it after, but Im guessing that will be after the dead line.
How long do I have?


From: Krzysztof Wicher <notifications@github.com>
Sent: Thursday, 14 November 2019 9:35 AM
To: dotnet/corefx <corefx@noreply.github.com>
Cc: Paul Farrand <paul.farrand@primetss.com.au>; Mention <mention@noreply.github.com>
Subject: Re: [dotnet/corefx] Fix System.IO.Ports.SerialPort GetPortNames() Returns garbage chars on Win10IoT Devices with USB Serial ports (#39943)


@PRIMETSS<https://github.com/PRIMETSS> I need to close the PR because of the repo merge (need to close or merge all PRs before 5PM today and this one has a merge conflict).

Do you think you'll have time to fix this? If so let me know if you need help with resurrecting this in the new repo

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/dotnet/corefx/pull/39943?email_source=notifications&email_token=ACWOOKUFMNQINTSUGPZ6JFLQTSB3FA5CNFSM4IINLCUKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOED77HNI#issuecomment-553644981>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACWOOKREBU3TKNBP42CMPZTQTSB3FANCNFSM4IINLCUA>.

krwq commented:
@PRIMETSS it's already after the deadline for corefx and this will need to be moved to dotnet/runtime repo once it becomes public. The directory structure will be the same so it should be straightforward to move the code. We don't have any specific ETA for when this will happen although I wouldn't expect more than a few weeks (optimistically much less but we need to do a lot of testing before we can make it public). You can still work in your branch and create a PR against master in your fork (make sure it's in sync with this master) and I can provide you feedback when you ping me.

krwq commented:
@PRIMETSS dotnet/runtime is now public
